src/transport_performance/population/rasterpop.py

```diff
@@ -0,0 +1,707 @@
+        """Round population data."""
+        self._xds = np.rint(self._xds)
+
+    def _threshold_population(self, threshold: int) -> None:
+        """Threshold population data."""
+        # `where()` is working differently to expectation here - it keeps
+        # values above the threshold and otherwise sets them to nan.
+        self._xds = self._xds.where(self._xds >= threshold)
+
+    def _to_geopandas(self, round: bool = False) -> None:
+        """Convert to geopandas dataframe."""
+        # vectorise to geopandas dataframe, setting data type to np.float32 -
+        # dtype is required by vecotrize function. Squeeze needed since shape
+        # is (1xnxm) (1 band in tiff file)
+        self.pop_gdf = vectorize(self._xds.squeeze(axis=0).astype(np.float32))
+
+        # dropna to remove nodata regions and those below threshold (if set)
+        self.pop_gdf = self.pop_gdf.dropna(subset=self.__var_name).reset_index(
+            drop=True
+        )
+
+        # if rounding, cast to int after removing na values
+        if round:
+            self.pop_gdf[self.__var_name] = self.pop_gdf[
+                self.__var_name
+            ].astype("int")
+
+        # add an id for each cell - needed for r5py
+        self.pop_gdf["id"] = np.arange(0, len(self.pop_gdf.index))
+
+        # re-order columns for consistency
+        self.pop_gdf = self.pop_gdf[["id", self.__var_name, "geometry"]]
+
+        # create centroid geodataframe using only necessary columns - save on
+        # duplicated data, can join gdfs on id if ever needed.
+        self.centroid_gdf = self.pop_gdf[["id", "geometry"]].copy()
+        self.centroid_gdf["centroid"] = self.centroid_gdf.geometry.centroid
+        self.centroid_gdf.set_geometry("centroid", inplace=True)
+        self.centroid_gdf.drop("geometry", axis=1, inplace=True)
+
+        # convert centroids to EPSG:4326 for use in r5py
+        if self.__crs != "EPSG:4326":
+            self.centroid_gdf = self.centroid_gdf.to_crs("EPSG:4326")
+
+    def _within_urban_centre(
+        self,
+        urban_centre_bounds: Type[Polygon],
+        urban_centre_crs: str = None,
+    ) -> None:
+        """Categorise grid whether within urban centre.
+
+        Parameters
+        ----------
+        urban_centre_bounds : Type[Polygon]
+            Polygon defining urban centre bounday
+        urban_centre_crs : str, optional
+            The urban centre polygon CRS, by default None meaning this is the
+            same CRS as the input raster data.
+
+        Raises
+        ------
+        TypeError
+            When `urban_centre_bounds` is not a shapely Polygon.
+
+        """
+        # defend against case where urban_centre_bounds is not a polygon
+        if not isinstance(urban_centre_bounds, Polygon):
+            raise TypeError(
+                f"Expected type {Polygon.__name__} for `urban_centre_bounds`, "
+                f"got {type(urban_centre_bounds).__name__}."
+            )
```

Comment:
Occurs to me that this snippet would be a useful utility function. Agreed to raise as technical debt issue.